go/mcap/ordered_lexer.go

```diff
@@ -0,0 +1,382 @@
+var ErrInsufficientSummary = errors.New("input does not contain required summary information")
+var ErrChunksNotIndexed = fmt.Errorf("%w: need chunk indexes", ErrInsufficientSummary)
+var ErrNoStatistics = fmt.Errorf("%w: need a statistics record", ErrInsufficientSummary)
+
+type recordIndex struct {
+	token     TokenType
+	timestamp uint64
+	offset    uint64
+	length    uint64
+}
+
+// OrderedLexer reads MCAP files record by record, yielding message records in ascending logTime
+// order.
+type OrderedLexer struct {
+	chunkContent   []byte
+	recordBuf      []byte
+	recordIndexes  []recordIndex
+	curRecordIndex int
+
+	chunkLoader chunkLoader
+
+	chunkTags            []chunkTag
+	curChunk             int
+	readyToYieldMessages bool
+	lexer                *Lexer
+}
+
+type chunkTag struct {
+	uncompressedSize uint64
+	mergeWithNext    bool
+}
+
+// tagChunksForMerging produces a chunkTag for every ChunkIndex in `info`. This tag determines
+// whether that chunk must be merged with one or more after it during reading.
+func tagChunksForMerging(info *Info) []chunkTag {
+	// ensure chunk indexes are sorted in physical order
+	sort.Slice(info.ChunkIndexes, func(i, j int) bool {
+		return info.ChunkIndexes[i].ChunkStartOffset < info.ChunkIndexes[j].ChunkStartOffset
+	})
+
+	chunkTags := make([]chunkTag, len(info.ChunkIndexes))
+	mergeUpTo := 0
+	for i, ci := range info.ChunkIndexes {
+		chunkTags[i].uncompressedSize = ci.UncompressedSize
+		// find the last chunk index that overlaps with this one, and ensure that this chunk
+		// and all between are merged into it.
+		for j := i + 1; j < len(info.ChunkIndexes); j++ {
+			if info.ChunkIndexes[j].MessageStartTime < ci.MessageEndTime {
+				if j > mergeUpTo {
+					mergeUpTo = j
+				}
+			}
+		}
+		if i < mergeUpTo {
+			chunkTags[i].mergeWithNext = true
+		}
+	}
+	return chunkTags
+}
+
+// getMaxRecordSize determines the largest single record which will be read from this MCAP during
+// lexing. This is used to pre-allocate the record buffer that the lexer reads into.
+func getMaxRecordSize(info *Info) uint64 {
+	// assume chunks are the largest records that we'll have to buffer.
+	realChunkSize := func(chunkIndex *ChunkIndex) uint64 {
+		// https://dev/spec#chunk-op0x06
+		return 8 + 8 + 8 + 4 + 4 + uint64(len(chunkIndex.Compression)) + 8 + chunkIndex.CompressedSize
+	}
+	var maxChunkRecordSize uint64
+	for _, chunkIndex := range info.ChunkIndexes {
+		size := realChunkSize(chunkIndex)
+		if size > maxChunkRecordSize {
+			maxChunkRecordSize = size
+		}
+	}
+	return maxChunkRecordSize
+}
+
+// NewOrderedLexer constructs a reader. It inspects the summary section of the MCAP, checking for
+// overlapping chunks. For any range of chunks that overlap in time range, they are tagged as
+// needing to be merged together in order to be sorted. It then determines the size of chunk buffer
+// required, and returns an error if this is greater than `maxBufferSize`.
+func NewOrderedLexer(
```

Comment:
I kinda wonder if this is too smart to call a lexer? The current structuring of the lib is that the lexer is the dumb low-level in-order iterator, and the reader is the index-aware smart iterator. It kinda seems like having a lexer construct a reader subverts that.

Reply:
Yeah, i really don't have a good name for this thing. I don't really intend to merge this - the point was to demonstrate that it passes conformance tests, then port the good parts into the existing seeking reader.